indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -269,11 +302,90 @@ public void setConf(Configuration config)
     }
   }
 
-  public static class IndexGeneratorReducer extends Reducer<BytesWritable, Writable, BytesWritable, Text>
+  public static class IndexGeneratorCombiner extends Reducer<BytesWritable, BytesWritable, BytesWritable, BytesWritable>
+  {
+    protected HadoopDruidIndexerConfig config;
+    private AggregatorFactory[] combiningAggs;
+
+    @Override
+    protected void setup(Context context)
+        throws IOException, InterruptedException
+    {
+      config = HadoopDruidIndexerConfig.fromConfiguration(context.getConfiguration());
+
+      List<AggregatorFactory> combiningAggsList = Lists.newArrayList();
+      for (AggregatorFactory factory : config.getSchema().getDataSchema().getAggregators()) {
+        combiningAggsList.add(factory.getCombiningFactory());
+      }
+      combiningAggs = combiningAggsList.toArray(new AggregatorFactory[0]);
+    }
+
+    @Override
+    protected void reduce(
+        final BytesWritable key, Iterable<BytesWritable> values, final Context context
+    ) throws IOException, InterruptedException
+    {
+
+      Iterator<BytesWritable> iter = values.iterator();
+      BytesWritable first = iter.next();
+
+      if(iter.hasNext()) {
+        SortableBytes keyBytes = SortableBytes.fromBytesWritable(key);
+        Bucket bucket = Bucket.fromGroupKey(keyBytes.getGroupKey()).lhs;
+
+        final int maxTotalBufferSize = config.getSchema().getTuningConfig().getBufferSize();
+        final int aggregationBufferSize = (int) ((double) maxTotalBufferSize
+                                                 * config.getSchema().getTuningConfig().getAggregationBufferRatio());
+
+        final StupidPool<ByteBuffer> bufferPool = new OffheapBufferPool(aggregationBufferSize);
+        IncrementalIndex index = makeIncrementalIndex(bucket, combiningAggs, bufferPool, config);
+        index.add(InputRowHelper.fromBytes(first.getBytes(), config.getSchema().getDataSchema().getAggregators()));
+
+        while(iter.hasNext()) {
+          context.progress();
+          InputRow value = InputRowHelper.fromBytes(
+              iter.next().getBytes(),
+              config.getSchema().getDataSchema().getAggregators()
```

Comment:
technically combining aggs could be used, but I want to keep things clear here and pass same set of aggs for derserialization that were passed at the time of serialization. I will store the refs globally though and use that.